# Database-style Operations on Dataframes

**Name**: Corpuz, Micki Laurren B.

**Section**: CPE22S3

**Instructor**: Engr. Roman Richard

### About the data

In this notebook, we will using daily weather data that was taken from the [National Centers for Environmental Information (NCEI) API](https://www.ncdc.noaa.gov/cdo-web/webservices/v2) . The data collection notebook contains
the process that was followed to collect the data.

    
Note: The NCEI is part of the National Oceanic and Atmospheric Administration (NOAA) and, as you can see from the URL for the API, this resource was created when the
NCEI was called the NCDC. Should the URL for this resource change in the future, you can search for the NCEI weather API to find the updated one.

### Background on the data

Data meanings:

- <b>PRCP </b>: precipitation in millimeters
- <b>SNOW </b>: snowfall in millimeters
- <b>SNWD </b>: snow depth in millimeters
- <b>TMAX </b>: maximum daily temperature in Celsius
- <b>TMIN </b>: minimum daily temperature in Celsius
- <b>TOBS </b> : temperature at time of observation in Celsius
- <b> WESF </b> : water equivalent of snow in millimeters

### Setup

In [11]:
import pandas as pd
weather = pd.read_csv('nyc_weather_2018.csv')
weather.head()

attributes datatype                 date            station  value
0       ,,N,     PRCP  2018-01-01T00:00:00  GHCND:US1CTFR0039    0.0
1       ,,N,     PRCP  2018-01-01T00:00:00  GHCND:US1NJBG0015    0.0
2       ,,N,     SNOW  2018-01-01T00:00:00  GHCND:US1NJBG0015    0.0
3       ,,N,     PRCP  2018-01-01T00:00:00  GHCND:US1NJBG0017    0.0
4       ,,N,     SNOW  2018-01-01T00:00:00  GHCND:US1NJBG0017    0.0

### Querying DataFrames

The query() method is an easier way of filtering based on some criteria. For example, we can use it to find all entries where snow was recorded:

In [15]:
snowdata = weather.query('datatype == "SNOW" and value > 0')
snowdata.head(5)

attributes datatype                 date            station  value
124       ,,N,     SNOW  2018-01-01T00:00:00  GHCND:US1NYWC0019   25.0
723       ,,N,     SNOW  2018-01-04T00:00:00  GHCND:US1NJBG0015  229.0
726       ,,N,     SNOW  2018-01-04T00:00:00  GHCND:US1NJBG0017   10.0
730       ,,N,     SNOW  2018-01-04T00:00:00  GHCND:US1NJBG0018   46.0
737       ,,N,     SNOW  2018-01-04T00:00:00  GHCND:US1NJES0018   10.0

This is equivalent to quering the data/weather.db SQLite database for SELECT * FROM weather WHERE datatype == "SNOW" AND value > 0 :

In [18]:
import sqlite3

with sqlite3.connect('weather.db') as connection:
    snow_data_from_db = pd.read_sql(
        'SELECT * FROM weather WHERE datatype == "SNOW" AND value > 0',
        connection
    )
    
snowdata.reset_index().drop(columns='index').equals(snow_data_from_db)

True

Note this is also equivalent to creating Boolean masks:

In [21]:
weather[(weather.datatype == 'SNOW') & (weather.value > 0)].equals(snowdata)

True

### Merging DataFrames

We have data for many different stations each day; however, we don't know what the stations are just their IDs. We can join the data in the data/weather_stations.csv
file which contains information from the stations endpoint of the NCEI API. Consult the weather_data_collection.ipynb notebook to see how this was collected. It looks like this:

In [25]:
station_info = pd.read_csv('weather_stations.csv')
station_info.head()

id                           name  latitude  longitude  \
0  GHCND:US1CTFR0022        STAMFORD 2.6 SSW, CT US   41.0641   -73.5770   
1  GHCND:US1CTFR0039          STAMFORD 4.2 S, CT US   41.0378   -73.5682   
2  GHCND:US1NJBG0001      BERGENFIELD 0.3 SW, NJ US   40.9213   -74.0020   
3  GHCND:US1NJBG0002  SADDLE BROOK TWP 0.6 E, NJ US   40.9027   -74.0834   
4  GHCND:US1NJBG0003           TENAFLY 1.3 W, NJ US   40.9147   -73.9775   

   elevation  
0       36.6  
1        6.4  
2       20.1  
3       16.8  
4       21.6

As a reminder, the weather data looks like this:

In [30]:
weather.head()

attributes datatype                 date            station  value
0       ,,N,     PRCP  2018-01-01T00:00:00  GHCND:US1CTFR0039    0.0
1       ,,N,     PRCP  2018-01-01T00:00:00  GHCND:US1NJBG0015    0.0
2       ,,N,     SNOW  2018-01-01T00:00:00  GHCND:US1NJBG0015    0.0
3       ,,N,     PRCP  2018-01-01T00:00:00  GHCND:US1NJBG0017    0.0
4       ,,N,     SNOW  2018-01-01T00:00:00  GHCND:US1NJBG0017    0.0

We can join our data by matching up the station_info.id column with the weather.station column. Before doing that though, let's see how many unique values we have:

In [33]:
station_info.id.describe()

count                   262
unique                  262
top       GHCND:US1CTFR0022
freq                      1
Name: id, dtype: object

While station_info has one row per station, the weather dataframe has many entries per station. Notice it also has fewer uniques:

In [36]:
weather.station.describe()

count                 80256
unique                  109
top       GHCND:USW00094789
freq                   4270
Name: station, dtype: object

When working with joins, it is important to keep an eye on the row count. Some join types will lead to data loss:

In [39]:
station_info.shape[0], weather.shape[0]

(262, 80256)

In [41]:
def grc(*dfs):
  return [df.shape[0] for df in dfs]
grc(station_info, weather)

[262, 80256]

The map() function is more efficient than list comprehensions. We can couple this with getattr() to grab any attribute for multiple dataframes

In [44]:
def getinf(attr, *dfs):
  return list(map(lambda x: getattr(x, attr), dfs))
getinf('shape', station_info,weather)

[(262, 5), (80256, 5)]

By default merge() performs an inner join. We simply specify the columns to use for the join. The left dataframe is the one we call merge() on, and the right one is passed in as an argument:

In [47]:
injoin = weather.merge(station_info,left_on='station', right_on='id')
injoin.sample(5, random_state=0)

attributes datatype                 date            station   value  \
27422       ,,W,     WDF5  2018-04-29T00:00:00  GHCND:USW00094741   310.0   
19317       ,,W,     WSF5  2018-03-24T00:00:00  GHCND:USW00094728     8.5   
13778       ,,W,     PGTM  2018-03-01T00:00:00  GHCND:USW00054743  2351.0   
39633       ,,W,     WSF2  2018-06-25T00:00:00  GHCND:USW00094789    11.2   
51025   ,,7,0700     TMAX  2018-08-19T00:00:00  GHCND:USC00280907    30.0   

                      id                              name  latitude  \
27422  GHCND:USW00094741          TETERBORO AIRPORT, NJ US  40.85000   
19317  GHCND:USW00094728       NY CITY CENTRAL PARK, NY US  40.77898   
13778  GHCND:USW00054743  CALDWELL ESSEX CO AIRPORT, NJ US  40.87639   
39633  GHCND:USW00094789  JFK INTERNATIONAL AIRPORT, NY US  40.63860   
51025  GHCND:USC00280907               BOONTON 1 SE, NJ US  40.89174   

       longitude  elevation  
27422  -74.06139        2.7  
19317  -73.96925       42.7  
13778  -74.28306       52.7  
39633  -73.76220        3.4  
51025  -74.39635       85.3

We can remove the duplication of information in the station and id columns by renaming one of them before the merge and then simply using on :

In [50]:
weather.merge(station_info.rename(dict(id='station'),axis=1),on='station').sample(5, random_state=0)

attributes datatype                 date            station   value  \
27422       ,,W,     WDF5  2018-04-29T00:00:00  GHCND:USW00094741   310.0   
19317       ,,W,     WSF5  2018-03-24T00:00:00  GHCND:USW00094728     8.5   
13778       ,,W,     PGTM  2018-03-01T00:00:00  GHCND:USW00054743  2351.0   
39633       ,,W,     WSF2  2018-06-25T00:00:00  GHCND:USW00094789    11.2   
51025   ,,7,0700     TMAX  2018-08-19T00:00:00  GHCND:USC00280907    30.0   

                                   name  latitude  longitude  elevation  
27422          TETERBORO AIRPORT, NJ US  40.85000  -74.06139        2.7  
19317       NY CITY CENTRAL PARK, NY US  40.77898  -73.96925       42.7  
13778  CALDWELL ESSEX CO AIRPORT, NJ US  40.87639  -74.28306       52.7  
39633  JFK INTERNATIONAL AIRPORT, NY US  40.63860  -73.76220        3.4  
51025               BOONTON 1 SE, NJ US  40.89174  -74.39635       85.3

We are losing stations that don't have weather observations associated with them, if we don't want to lose these rows, we perform a right or left join instead of the inner join:

In [53]:
left_join = station_info.merge(weather, left_on='id',right_on='station', how='left')
right_join = weather.merge(station_info, left_on='station',right_on='id', how='right')

right_join.tail()

attributes datatype                 date            station  value  \
80404       ,,W,     WDF5  2018-12-31T00:00:00  GHCND:USW00094789  130.0   
80405       ,,W,     WSF2  2018-12-31T00:00:00  GHCND:USW00094789    9.8   
80406       ,,W,     WSF5  2018-12-31T00:00:00  GHCND:USW00094789   12.5   
80407       ,,W,     WT01  2018-12-31T00:00:00  GHCND:USW00094789    1.0   
80408       ,,W,     WT02  2018-12-31T00:00:00  GHCND:USW00094789    1.0   

                      id                              name  latitude  \
80404  GHCND:USW00094789  JFK INTERNATIONAL AIRPORT, NY US   40.6386   
80405  GHCND:USW00094789  JFK INTERNATIONAL AIRPORT, NY US   40.6386   
80406  GHCND:USW00094789  JFK INTERNATIONAL AIRPORT, NY US   40.6386   
80407  GHCND:USW00094789  JFK INTERNATIONAL AIRPORT, NY US   40.6386   
80408  GHCND:USW00094789  JFK INTERNATIONAL AIRPORT, NY US   40.6386   

       longitude  elevation  
80404   -73.7622        3.4  
80405   -73.7622        3.4  
80406   -73.7622        3.4  
80407   -73.7622        3.4  
80408   -73.7622        3.4

In [55]:
left_join.sort_index(axis=1).sort_values(['date','station']).reset_index().drop(columns='index').equals(
    right_join.sort_index(axis=1).sort_values(['date','station']).reset_index().drop(columns='index')
)
  

True

Note we have additional rows in the left and right joins because we kept all the stations that didn't have weather observations:

In [58]:
getinf('shape', injoin, left_join, right_join)

[(80256, 10), (80409, 10), (80409, 10)]

If we query the station information for stations that have NY in their name, believing that to be all the stations that record weather data for NYC and perform an outer join, we can see where the mismatches occur:

In [61]:
outer_join = weather.merge(
    station_info[station_info.name.str.contains('NY')],
    left_on='station',right_on='id', how='outer', indicator=True
)
pd.concat([outer_join.sample(4, random_state=0),outer_join[outer_join.station.isna()].head(2)]) # I used concat and not used append as it was an older version of pandas

attributes datatype                 date            station   value  \
17259       ,,N,     SNOW  2018-07-20T00:00:00  GHCND:US1NJMS0075     0.0   
76178       ,,W,     AWND  2018-01-12T00:00:00  GHCND:USW00094789     7.2   
73410  T,,W,2400     PRCP  2018-03-16T00:00:00  GHCND:USW00094745     0.0   
74822       ,,W,     PGTM  2018-08-20T00:00:00  GHCND:USW00094745  1641.0   
14954        NaN      NaN                  NaN                NaN     NaN   
14955        NaN      NaN                  NaN                NaN     NaN   

                      id                                  name  latitude  \
17259                NaN                                   NaN       NaN   
76178  GHCND:USW00094789      JFK INTERNATIONAL AIRPORT, NY US  40.63860   
73410  GHCND:USW00094745         WESTCHESTER CO AIRPORT, NY US  41.06694   
74822  GHCND:USW00094745         WESTCHESTER CO AIRPORT, NY US  41.06694   
14954  GHCND:US1NJMS0036  PARSIPPANY TROY HILLS TWP 2.1, NJ US  40.86560   
14955  GHCND:US1NJMS0039  PARSIPPANY TROY HILLS TWP 1.3, NJ US  40.85330   

       longitude  elevation      _merge  
17259        NaN        NaN   left_only  
76178   -73.7622        3.4        both  
73410   -73.7075      115.5        both  
74822   -73.7075      115.5        both  
14954   -74.3851       64.3  right_only  
14955   -74.4470       94.2  right_only

These joins are equivalent to their SQL counterparts. Below is the inner join. Note that to use equals() you will have to do some manipulation of the dataframes to line them up:

In [64]:
import sqlite3 as sq3

with sq3.connect('weather.db') as connection:
  inner_join_from_db = pd.read_sql('SELECT * FROM weather JOIN stations ON weather.station == stations.id',connection)

inner_join_from_db.shape == injoin.shape

True

Revisit the dirty data from the previous module.

In [67]:
dirty_data = pd.read_csv('dirty_data2.csv', index_col='date').drop_duplicates().drop(columns='SNWD')
dirty_data.head()

station  PRCP   SNOW    TMAX  TMIN  TOBS  WESF  \
date                                                                            
2018-01-01T00:00:00                  ?   0.0    0.0  5505.0 -40.0   NaN   NaN   
2018-01-02T00:00:00  GHCND:USC00280907   0.0    0.0    -8.3 -16.1 -12.2   NaN   
2018-01-03T00:00:00  GHCND:USC00280907   0.0    0.0    -4.4 -13.9 -13.3   NaN   
2018-01-04T00:00:00                  ?  20.6  229.0  5505.0 -40.0   NaN  19.3   
2018-01-05T00:00:00                  ?   0.3    NaN  5505.0 -40.0   NaN   NaN   

                    inclement_weather  
date                                   
2018-01-01T00:00:00               NaN  
2018-01-02T00:00:00             False  
2018-01-03T00:00:00             False  
2018-01-04T00:00:00              True  
2018-01-05T00:00:00               NaN

We need to create two dataframes for the join. We will drop some unecessary columns as well for easier viewing:

In [70]:
valid_station = dirty_data.query('station != "?"').copy().drop(columns=['WESF','station'])
sta_with_wesf = dirty_data.query('station == "?"').copy().drop(columns=['station', 'TOBS', 'TMIN', 'TMAX'])

Our column for the join is the index in both dataframes, so we must specify left_index and right_index :

In [73]:
valid_station.merge(sta_with_wesf, left_index=True, right_index=True).query('WESF>0').head()

PRCP_x  SNOW_x  TMAX  TMIN  TOBS inclement_weather_x  \
date                                                                        
2018-01-30T00:00:00     0.0     0.0   6.7  -1.7  -0.6               False   
2018-03-08T00:00:00    48.8     NaN   1.1  -0.6   1.1               False   
2018-03-13T00:00:00     4.1    51.0   5.6  -3.9   0.0                True   
2018-03-21T00:00:00     0.0     0.0   2.8  -2.8   0.6               False   
2018-04-02T00:00:00     9.1   127.0  12.8  -1.1  -1.1                True   

                     PRCP_y  SNOW_y  WESF inclement_weather_y  
date                                                           
2018-01-30T00:00:00     1.5    13.0   1.8                True  
2018-03-08T00:00:00    28.4     NaN  28.7                 NaN  
2018-03-13T00:00:00     3.0    13.0   3.0                True  
2018-03-21T00:00:00     6.6   114.0   8.6                True  
2018-04-02T00:00:00    14.0   152.0  15.2                True

The columns that existed in both dataframes, but didn't form part of the join got suffixes added to their names: _ x _ for columns from the left dataframe and _y_ for columns from the right dataframe. We can customize this with the suffixes argument:

In [76]:
valid_station.merge(sta_with_wesf, left_index=True, right_index=True, suffixes = ('','_?')).query('WESF>0').head()

PRCP   SNOW  TMAX  TMIN  TOBS inclement_weather  PRCP_?  \
date                                                                           
2018-01-30T00:00:00   0.0    0.0   6.7  -1.7  -0.6             False     1.5   
2018-03-08T00:00:00  48.8    NaN   1.1  -0.6   1.1             False    28.4   
2018-03-13T00:00:00   4.1   51.0   5.6  -3.9   0.0              True     3.0   
2018-03-21T00:00:00   0.0    0.0   2.8  -2.8   0.6             False     6.6   
2018-04-02T00:00:00   9.1  127.0  12.8  -1.1  -1.1              True    14.0   

                     SNOW_?  WESF inclement_weather_?  
date                                                   
2018-01-30T00:00:00    13.0   1.8                True  
2018-03-08T00:00:00     NaN  28.7                 NaN  
2018-03-13T00:00:00    13.0   3.0                True  
2018-03-21T00:00:00   114.0   8.6                True  
2018-04-02T00:00:00   152.0  15.2                True

Since we are joining on the index, an easier way is to use the join() method instead of merge() . Note that the suffix parameter is now lsuffix for the left dataframe's suffix and rsuffix for the right one's:

In [79]:
valid_station.join(sta_with_wesf, rsuffix='_?').query('WESF >0').head()

PRCP   SNOW  TMAX  TMIN  TOBS inclement_weather  PRCP_?  \
date                                                                           
2018-01-30T00:00:00   0.0    0.0   6.7  -1.7  -0.6             False     1.5   
2018-03-08T00:00:00  48.8    NaN   1.1  -0.6   1.1             False    28.4   
2018-03-13T00:00:00   4.1   51.0   5.6  -3.9   0.0              True     3.0   
2018-03-21T00:00:00   0.0    0.0   2.8  -2.8   0.6             False     6.6   
2018-04-02T00:00:00   9.1  127.0  12.8  -1.1  -1.1              True    14.0   

                     SNOW_?  WESF inclement_weather_?  
date                                                   
2018-01-30T00:00:00    13.0   1.8                True  
2018-03-08T00:00:00     NaN  28.7                 NaN  
2018-03-13T00:00:00    13.0   3.0                True  
2018-03-21T00:00:00   114.0   8.6                True  
2018-04-02T00:00:00   152.0  15.2                True

Joins can be very resource-intensive, so it's a good idea to figure out what type of join you need using set operations before trying the join itself. The pandas set operations are performed on the index, so whichever columns we will be joining on will need to be the index. Let's go back to the weather and station_info dataframes and set the station ID columns as the index:

In [82]:
weather.set_index('station', inplace=True)
station_info.set_index('id', inplace=True)

The intersection will tell us the stations that are present in both dataframes. The result will be the index when performing an inner join:

In [85]:
weather.index.intersection(station_info.index)

Index(['GHCND:US1CTFR0039', 'GHCND:US1NJBG0015', 'GHCND:US1NJBG0017',
       'GHCND:US1NJBG0018', 'GHCND:US1NJBG0023', 'GHCND:US1NJBG0030',
       'GHCND:US1NJBG0039', 'GHCND:US1NJBG0044', 'GHCND:US1NJES0018',
       'GHCND:US1NJES0024',
       ...
       'GHCND:US1NJMS0047', 'GHCND:US1NYSF0083', 'GHCND:US1NYNY0074',
       'GHCND:US1NJPS0018', 'GHCND:US1NJBG0037', 'GHCND:USC00284987',
       'GHCND:US1NJES0031', 'GHCND:US1NJMD0086', 'GHCND:US1NJMS0097',
       'GHCND:US1NJMN0081'],
      dtype='object', length=109)

In [87]:
weather.index.difference(station_info.index)

Index([], dtype='object')

We lose 114 stations from the station_info dataframe, however:

In [90]:
station_info.index.difference(weather.index)

Index(['GHCND:US1CTFR0022', 'GHCND:US1NJBG0001', 'GHCND:US1NJBG0002',
       'GHCND:US1NJBG0005', 'GHCND:US1NJBG0006', 'GHCND:US1NJBG0008',
       'GHCND:US1NJBG0011', 'GHCND:US1NJBG0012', 'GHCND:US1NJBG0013',
       'GHCND:US1NJBG0020',
       ...
       'GHCND:USC00308322', 'GHCND:USC00308749', 'GHCND:USC00308946',
       'GHCND:USC00309117', 'GHCND:USC00309270', 'GHCND:USC00309400',
       'GHCND:USC00309466', 'GHCND:USC00309576', 'GHCND:USW00014708',
       'GHCND:USW00014786'],
      dtype='object', length=153)

The symmetric difference will tell us what gets lost from both sides. It is the combination of the set difference in both directions: 216 + 114 = 330 which was station_info shape was

In [93]:
ny_in_name = station_info[station_info.name.str.contains('NY')]

ny_in_name.index.difference(weather.index).shape[0]\
+ weather.index.difference(ny_in_name.index).shape[0]\
== weather.index.symmetric_difference(ny_in_name.index).shape[0]

True

The union will show us everything that will be present after a full outer join. Note that since these are sets (which don't allow duplicates by definition), we must pass unique
entries for union:

In [96]:
weather.index.unique().union(station_info.index)

Index(['GHCND:US1CTFR0022', 'GHCND:US1CTFR0039', 'GHCND:US1NJBG0001',
       'GHCND:US1NJBG0002', 'GHCND:US1NJBG0003', 'GHCND:US1NJBG0005',
       'GHCND:US1NJBG0006', 'GHCND:US1NJBG0008', 'GHCND:US1NJBG0010',
       'GHCND:US1NJBG0011',
       ...
       'GHCND:USW00014708', 'GHCND:USW00014732', 'GHCND:USW00014734',
       'GHCND:USW00014786', 'GHCND:USW00054743', 'GHCND:USW00054787',
       'GHCND:USW00094728', 'GHCND:USW00094741', 'GHCND:USW00094745',
       'GHCND:USW00094789'],
      dtype='object', length=262)

Note that the symmetric difference is actually the union of the set differences:

In [99]:
ny_in_name = station_info[station_info.name.str.contains('NY')]
ny_in_name.index.difference(weather.index).union(weather.index.difference(ny_in_name.index)).equals(
weather.index.symmetric_difference(ny_in_name.index)
)

True